### DISCUSSION TITLE
Using a RAG agent or any other agent implementation with a phased coordinator

### DISCUSSION THREAD

**Discussion by fjuma** [wise-agents/wise-agents#266] [Ideas]
Currently, a phased coordinator needs to be used in conjunction with a `CollaboratorWiseAgent` to ensure that when processing a request, we make use of the chat history and when sending the response back to the coordinator, the response is added to the shared context and just an ACK message is sent back to the coordinator.

That means that it's currently not possible to make use of something like a `RAGWiseAgent` (or any of our other agent implementations) in conjunction with the phased coordinator because these implementations currently don't process requests in the same way a `CollaboratorWiseAgent` does.

We can't just update these implementations to always make use of the chat history and add the response to the shared context because we don't know if the user will want to use the agent in combination with a sequential coordinator or a phased coordinator or standalone agent, etc.

For the sequential coordinator case, right now, it is possible to make use of our agent implementations there because this coordinator implementation just assumes that each agent will send its response back to the sender. When we modify or add a new sequential coordinator that sends the requests/responses from one agent to another directly without going back to the coordinator after each agent (https://github.com/wise-agents/wise-agents/issues/238), we'll have a similar issue as well, i.e., the existing agent implementations won't work with this type of coordinator because we'll require special handling for sending the response and ACK.

One option is to make it possible for a user to wrap an agent implementation with a `CollaboratorWiseAgent`. When processing a request, we could then ensure that the chat history is retrieved if required and that we take the required steps when sending a response. This might require us to revisit our `WiseAgent` `process_request` method and have it call a few methods so that we can split out the different parts that could differ between coordinator implementations (whether or not to retrieve chat history, actually processing the request with an LLM (or without an LLM), and sending the response back) to make it possible to use an agent implementation with different types of coordinators. Am looking closer into what this would actually look like.

Thoughts?

**Comment by maeste** [comment]
> One option is to make it possible for a user to wrap an agent implementation with a `CollaboratorWiseAgent`. When processing a 

What do you mean with "wrap" here? Adding another agent that call user agent? It' is certainly a possibility, but it would make the config, rout and debug very hard.
What about writing a decorator (an annotation) for process_request and other process_* to ensure the shared memory stuff is guarantee working with WiseMessage input/output. It's just an idea I haven't digged in the code to see if it makes sense and if it is possible

Best,
S.

**Reply by fjuma** [reply]
> > One option is to make it possible for a user to wrap an agent implementation with a `CollaboratorWiseAgent`. When processing a
> 
> What do you mean with "wrap" here? Adding another agent that call user agent?

My initial thought was that a CollaboratorAgent could have a delegate agent and so the collaborator would be responsible for calling the delegate agent with the stuff passed from shared memory and it would handle the response as well.

>What about writing a decorator (an annotation) for process_request and other process_* to ensure the shared memory stuff is guarantee working with WiseMessage input/output. It's just an idea I haven't digged in the code to see if it makes sense and if it is possible

I also started looking into that option yesterday after starting this discussion :)

Am looking more into this now.

**Comment by fjuma** [comment]
Created an issue to track this: https://github.com/wise-agents/wise-agents/issues/281
